sdk/core/core/src/az_http_request_builder.c

```diff
@@ -3,48 +3,188 @@
+      : NULL;
+}
+
 az_result az_http_request_builder_init(
     az_http_request_builder * const p_hrb,
     az_span const buffer,
-    int16_t const max_url_size,
     az_const_span const method_verb,
-    az_const_span const initial_url) {
-  (void)p_hrb;
-  (void)buffer;
-  (void)max_url_size;
-  (void)method_verb;
-  (void)initial_url;
+    az_const_span const initial_url,
+    uint16_t const max_url_size,
+    uint16_t const max_headers) {
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
+
+  if (!az_span_is_valid(buffer) || !az_const_span_is_valid(method_verb)
+      || !az_const_span_is_valid(initial_url) || max_url_size < initial_url.size) {
+    return AZ_ERROR_ARG;
+  }
+
+  if (buffer.size < max_url_size
+      || (max_headers > 0 && get_headers_start(buffer, max_url_size, max_headers) == NULL)) {
+    return AZ_ERROR_BUFFER_OVERFLOW;
+  }
+
+  AZ_RETURN_IF_FAILED(az_span_set(buffer, '\0'));
+
+  az_span uri_buf = { 0, 0 };
+  AZ_RETURN_IF_FAILED(az_span_copy(
+      (az_span){ .begin = buffer.begin, .size = max_url_size }, initial_url, &uri_buf));
+
+  *p_hrb = (az_http_request_builder){ .buffer = buffer,
+                                      .method_verb = method_verb,
+                                      .url = uri_buf,
+                                      .max_url_size = max_url_size,
+                                      .max_headers = max_headers,
+                                      .retry_headers_start = max_headers,
+                                      .headers_end = 0 };
+
   return AZ_OK;
 }
 
 az_result az_http_request_builder_set_query_parameter(
     az_http_request_builder * const p_hrb,
     az_const_span const name,
     az_const_span const value) {
-  (void)p_hrb;
-  (void)name;
-  (void)value;
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
+
+  if (!az_const_span_is_valid(name) || !az_const_span_is_valid(value) || name.size == 0) {
+    return AZ_ERROR_ARG;
+  }
+
+  size_t new_url_size;
+  bool const has_value = value.size > 0;
+  {
+    az_span new_url_span = { .begin = p_hrb->url.begin };
+    size_t const extra_chars_size = sizeof((uint8_t)'?') + (has_value ? sizeof((uint8_t)'=') : 0);
+    size_t const name_and_value_size = name.size + value.size;
+    size_t const appended_size = name_and_value_size + extra_chars_size;
+
+    new_url_size = p_hrb->url.size + appended_size;
+
+    if (name_and_value_size < name.size || name_and_value_size < value.size
+        || appended_size < name_and_value_size || appended_size < extra_chars_size
+        || new_url_size < appended_size || new_url_size < p_hrb->url.size
+        || new_url_size > p_hrb->max_url_size) {
+      return AZ_ERROR_BUFFER_OVERFLOW;
+    }
+
+    new_url_span.size = new_url_size;
+
+    if (az_const_span_is_overlap(az_span_to_const_span(new_url_span), name)
+        || az_const_span_is_overlap(az_span_to_const_span(new_url_span), value)) {
+      return AZ_ERROR_ARG;
+    }
+  }
+
+  bool first_parameter = true;
+  for (size_t i = p_hrb->url.size; i > 0;) {
```

Comment:
what about adding an extra field `has_query` into control block (builder) so we don't need to loop all url on every `add_query` ??

Reply:
We can; we will need to scan the initial URL in the init method. Let's discuss with the team.

Reply:
Yes, this is what I was saying above. It's a nice perf improvement.

Reply:
I'm not changing it in this PR. Let's discuss. I don't think we should optimize URL formatting further, I explained in other comment.